StreamChatClient_v3/Controllers/ChangeAggregator.swift

```diff
@@ -0,0 +1,113 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+
+/// This enum describes the changes of the given collections of items.
+public enum Change<Item> {
+    /// A new item was inserted on the given index path.
+    case insert(_ item: Item, index: IndexPath)
+    
+    /// An item was moved from `fromIndex` to `toIndex`. Moving an item also automatically generates
+    /// an `update` change.
+    case move(_ item: Item, fromIndex: IndexPath, toIndex: IndexPath)
+    
+    /// An item was updated at the given `index`. An `update` change is also automatically generated by moving an item.
+    case update(_ item: Item, index: IndexPath)
+    
+    /// An item was removed from the given `index`.
+    case remove(_ item: Item, index: IndexPath)
+}
+
+extension Change: Equatable where Item: Equatable {}
+
+/// When this object is set as `NSFetchedResultsControllerDelegate`, it aggregates the callbacks from the fetched results
+/// controller and forwards them in the way of `[Change<Item>]`. You can set the `onChange` callback to receive these updates.
+class ChangeAggregator<DTO: NSManagedObject, Item>: NSObject, NSFetchedResultsControllerDelegate {
+    // TODO: Extend this to also provide `CollectionDifference` and `NSDiffableDataSourceSnapshot`
+    
+    /// Used for converting the `DTO`s provided by `FetchResultsController` to the resulting `Item`.
+    let itemCreator: (DTO) -> Item?
+    
+    /// The current observer of the aggregator.
+    private(set) var onChangeCallback: (([Change<Item>]) -> Void)?
+    
+    /// An array of changes in the current update. It gets reset every time `controllerWillChangeContent` is called, and
+    /// published to the observer when `controllerDidChangeContent` is called.
+    private var currentChanges: [Change<Item>] = []
+    
+    /// Creates a new `ChangeAggregator`.
+    ///
+    /// - Parameter itemCreator: Used for converting the `NSManagedObject`s provided by `FetchResultsController`
+    /// to the resulting `Item`.
+    init(itemCreator: @escaping (DTO) -> Item?) {
+        self.itemCreator = itemCreator
+    }
+    
+    /// Adds an observer for the aggregator.
+    ///
+    /// - Parameters:
+    ///   -  callback: Called with the aggregated changes after `FetchResultsController` calls
+    ///     `controllerDidChangeContent` on its delegate.
+    ///   - changes: The aggregated changes reported by `FetchResultsController`.
+    func onChange(_ callback: @escaping (_ changes: [Change<Item>]) -> Void) {
+        onChangeCallback = callback
+    }
```

Comment:
Why is this a function, instead of just assigning to `onChangeCallback`?

Reply:
Haha, no idea :)

Reply:
Fixed in 1dc1219fa82189419dc667a31a4ed25affe28caf 7f1c18c2b42eb96b17cf1fdb109c6673aa2062bb